src/Security/User/ContaoUserProvider.php

```diff
@@ -0,0 +1,105 @@
+<?php
+
+declare(strict_types=1);
+
+/*
+ * This file is part of Contao.
+ *
+ * Copyright (c) 2005-2017 Leo Feyer
+ *
+ * @license LGPL-3.0+
+ */
+
+namespace Contao\CoreBundle\Security\User;
+
+use Contao\CoreBundle\Framework\ContaoFrameworkInterface;
+use Contao\User;
+use Symfony\Component\Security\Core\Exception\UnsupportedUserException;
+use Symfony\Component\Security\Core\Exception\UsernameNotFoundException;
+use Symfony\Component\Security\Core\User\UserInterface;
+use Symfony\Component\Security\Core\User\UserProviderInterface;
+
+class ContaoUserProvider implements UserProviderInterface
+{
+    /**
+     * @var ContaoFrameworkInterface
+     */
+    private $framework;
+
+    /**
+     * @var string
+     */
+    private $userClass;
+
+    /**
+     * @param ContaoFrameworkInterface $framework
+     * @param string                   $userClass
+     */
+    public function __construct(ContaoFrameworkInterface $framework, string $userClass)
```

Comment:
I don't like this at all. There are only two types of Contao users and it should not be possible to provide anything but their two classes.

Having a back end and a front end user was fine – why did you change it?

Reply:
I don't get that as well.

Reply:
Because both classes contain exactly the same code except for the class to check for. Remember the discussion we had with the AssetContext? 😉

Reply:
Except you explicitly can have multiple asset contexts but you cannot have another user than back end user and front end user.

Code duplication can also be reduced by an abstract parent class or a trait.

Reply:
I remember having the exact same discussion in the assets pull request. We can do parent classes as well, but that does not make a difference as you could write a third child class the same way you could register a third service with a different argument. Both ways are technically perfectly fine, I just don't see the need to write three files instead of one and use our beloved dependency injection. But I can change that if you insist…

Reply:
I don't insist.